Readme renderer update

GuusVink opened the pull request:
Included general information of the repo
+
formed the readme template outside of the JS file (for ease of markdown editing)
+
included all types that are visited by the analyzer and thus included in the generated JSON (along with hyperrefs to the JavaParser documentation) -> this addresses the 'issue' of the current state of the analyzer (i.e., what language constructs are supported)

wandernauta commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+* [**Analyzer**](#analyzer): A tool to analyze Java projects.
+* [**Libraries**](#support-libraries): Assists in generating applications that can create plain text files such as MarkDown, AsciiDoc, PlantUML, Mermaid, and more.
+
+> **Note**: As this repository is currently association to a University assignment, it is unfortunately not possible to directly contribute to this project. 
```
Aww, are you sure about this? I would think we'd gladly take patches, if people have them.

GuusVink replied:
I am not sure about this as we are graded on how we work as a project group and the work we deliver should be our own. 
On one hand, it could be considered positive if we manage an open-source project correctly resulting in a better product for Info Support. On the other hand, other people should probably not take over our work.
But again, this is a guess from my side on how Petra would view this (maybe discuss this sometime with Petra and others?)

GuusVink replied:
I created a general section about this as discussed with Petra

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,57 @@
+const json = JSON.parse(fs.readFileSync(process.argv[2]));
 
+// Section for rendering the command line options
+const CMD_LINE_TABLE_TEMPLATE = `
```
Why is this here, and not in `README_template.md`?

GuusVink replied:
Because the Javascript code filling the table is dependent on the exact form of this smaller template (i.e. #columns & headers)

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,57 @@
+
+// Section rendering the implemented visited methods
+
+const JAVAPARSER_DOC_SITE = 'https://javadoc.io/static/com.github.javaparser/javaparser-core/3.25.6/';
```
Probably not a big difference, but we use version 3.25.1, not 3.25.6 (see analyzer-java's pom.xml).

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,57 @@
+
+const JAVAPARSER_DOC_SITE = 'https://javadoc.io/static/com.github.javaparser/javaparser-core/3.25.6/';
+function getMarkdownLink(type) {
+  let splitted = type.split('.');
```
Here and below, all except `finalReadme` can be `const` instead of `let`.

GuusVink replied:
I am not that familiar with the JS code styling guidelines, so excuse me for my ignorance (:  But why would `const` be preferred if it does not semantically represent a constant?
And if using `const`, would you also not need to change the naming to capital snake case?

wandernauta replied:
We didn't officially pick any JavaScript style guide, but if you want to be consistent with the existing code, I went with:

* mark all variables `const` where possible, to indicate we don't expect them to be reassigned
* give only "real" constants (in this case, hard-coded settings) shouty-case names

That happens to align with Google JavaScript style, see [here](https://google.github.io/styleguide/jsguide.html#features-use-const-and-let) and [here](https://google.github.io/styleguide/jsguide.html#naming-constant-names).

GuusVink replied:
I have changed this, for the report, it may be also nice to mention this style guide in the QA

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,57 @@
+// Combining everything
+
+try {
+  let readmeTemplate = fs.readFileSync('README_template.md', 'utf8');
```
Suggest extracting `'README_template.md'` this as a constant and putting it near line 11.

wandernauta reviewed (commented):
First round done, some general questions and some suggestions for the actual code.

GuusVink commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+## Getting started
+
+The process of rendering documentation can be summarized as follows:
+1. **Analysis**: Run the Analyzer over the source code of your project. This creates a LivingDocumentation JSON file.
```
The only disadvantage with that is that if the options change (either the literal spelling or usage) this would have to be changed manually as opposed to the table below.
Would be kind of undermining if as a consequence of this the readme would contain a contradiction somewhere in the future (the word undermining is maybe a bit extreme but you get the sentiment).

This is why I tried to keep the 'fixed parts' rather general. Curious how you see this

wandernauta replied:
I get what you mean. Indeed, let's keep it like this.

GuusVink commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+
+[~cmd-line-options]
+
+### Maven plugin
```
Would a note containing something like 'at the moment the plugin is not published, but it can be installed manually' be a good solution?

wandernauta replied:
Sure! We'd have to include instructions on how to do that manual installation, and make sure that those indeed work.

wandernauta replied:
That really works - just `mvn install` inside the plugin directory and off you go? That's neat.

I would have expected there to be more steps, to build and install the analyzer perhaps, but if you've tried it and confirmed that's not needed, even for people who haven't used this before, great.

GuusVink commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+The analyzer can also be used as a Maven plugin (which is defined in [ldj-maven-plugin](/ldj-maven-plugin)).
+For an example on how to use the Maven plugin, see the `pom.xml` configuration in [ldj-maven-plugin-example](/ldj-maven-plugin-example).
+
+### State of analyzer
```
Main reason was to address questions like _"What types of classes and relations can this tool model?"_. This gives a dynamic list of the java features that are covered by the analyzer. Relation to the JavaParser documentation is to give a description of the exact java feature (which is not generally deductible from our code I think). In the future this sort of documentation could be replaced by or linked to the JSON schema documentation; answering the above question.

wandernauta replied:
Hmm, right - I think I get what you mean, but if this is the intention, the list is not accurate, right? It includes `CatchClause,` but we do not include information about catch clauses in the JSON. I think we actually can not do so, without breaking compatibility; so listing it as something we already do seems odd to me.

If the hyperlinks are currently not useful, but will be useful in the future, I would suggest removing them for now and adding them at that future point.

GuusVink replied:
As talked about, right now this list is excluded. If in a later stadium it might be useful to again include this information (albeit in some other documentation artefact), we can revert (or copy from the inverse of ) commit `fff85fa0339fcb6100e516939d0fc3bac867e460` 

Furthermore, this readme now contains a sort of placeholder reference to the JSON schema documentation which is yet to be approved/merged

GuusVink replied:
Note for future reference:  commit `ee5deea9df4c85eab988b5098042fb2d5573cc77` now also contains a deletion of a required statement for this feature

GuusVink commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+can be identified (which are copied from [LivingDocumentation][ldoc] description):
+
+* [**Analyzer**](#analyzer): A tool to analyze Java projects.
+* [**Libraries**](#support-libraries): Assists in generating applications that can create plain text files such as MarkDown, AsciiDoc, PlantUML, Mermaid, and more.
```
I have now added a note that they are not included in the project yet. As they are relevant for understanding the entire project (LD in general) I have still included them in the general explanations

GuusVink commented:
@wandernauta I have addressed all issues. Once you approve (and merge) this PR, please either add a reference to the JSON schema documentation in readme section 'State of analyzer' (if at that moment the JSON schema doc is merged with main). Or add a note to the JSON schema doc PR to insert a reference to that doc within the README template of this PR

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,43 @@
-const json = JSON.parse(fs.readFileSync(process.argv[2]));
-fs.writeFileSync(OUTPUT, TEMPLATE.replace('~', () => render(json)));
+const cmdOptionsTable = CMD_LINE_TABLE_TEMPLATE.replace('~', () => renderCmdOptions(json));
+const visitor = json.filter((type) => type.FullName === 'com.infosupport.ldoc.analyzerj.AnalysisVisitor')[0];
```
This now seems unused.

GuusVink replied:
Correct, missed that one

wandernauta commented on ldj-readme-renderer/render.js:
```diff
@@ -8,45 +8,43 @@
+  const finalReadme = readmeTemplate.replace('[~cmd-line-options]', cmdOptionsTable);
+  fs.writeFileSync(OUTPUT, finalReadme);
+} catch (err) {
+  console.error(err);
```
I think the default behavior for an uncaught exception is to write the error to the console - does catching and logging it explicitly here make an important difference?

If not, I'd suggest removing the try/catch. That also makes it consistent with the `readFileSync` of the JSON file, above.

GuusVink replied:
For the purpose of this renderer does indeed not make a difference. Agreed, so adjusted

wandernauta reviewed (commented):
Did a second round, found some smaller things.

I'll keep an eye on #90 as well so we can link that as soon as it lands.

GuusVink commented on ldj-readme-renderer/README_template.md:
```diff
@@ -0,0 +1,66 @@
+## State of the project
+
+As of right now, this project is linked to a student assignment originating from the University of
+Twente and InfoSupport. The aim of the project is to make the [LivingDocumentation][ldoc] project
```
Weird. Fixed both

wandernauta reviewed (approved):
Looks good to me!
